### ISSUE TITLE
Sequelocity on .NET Core 

### ISSUE THREAD

**Issue by westdavidr** [AmbitEnergyLabs/Sequelocity.NET#25]
Is there a roadmap for porting Sequelocity to .NET Core?

**Comment by randyburden** [comment]
No, not currently for three reasons: 

1. I was waiting for .NET Core to officially 1.0 to officially launch [which it is now has as of June 27th](https://blogs.msdn.microsoft.com/dotnet/2016/06/27/announcing-net-core-1-0/).
2. I was waiting to see how the whole reflection API would shake out because it appeared that almost everything involving reflection would require changes and to support both .NET Core and .NET 4.6 would require adding compiler directives to support both framework versions. 
3. Last I heard is that MS is hard at work on bringing back everything they said wouldn't be included in Core which means that if we just wait then the standard reflection API will get implemented along with everything else and thus compatibility for .NET Core will just come for free as .NET Core reaches parity with the standard framework.

I haven't heard that any teams are eager to use .NET Core at Ambit but if you have a burning need we accept pull requests. ;)

If I get time I'll create a sample .NET core app and paste in Sequelocity to gauge how much work it's going to take.

**Comment by randyburden** [comment]
Okay so I created a .NET Core app, pasted in Sequelocity and it lit up like a red Christmas tree with errors! ;)

tldr; **This is not a trivial task.**

#### Resolvable Issues
- Installing System.Data.Common via NuGet
- Installing System.Data.SqlClient via NuGet
- Modifying ```type.``` to ```type.GetTypeInfo().``` although this was just to try to get it to build and wouldn't work as it's not compatible with the full version of the framework thus this is going to require compiler directives.
- AppDomain class doesn't exist although I should be able to work around that as all we are using it for is to get the host application name.
- Serializable class doesn't exist although that can be worked around by adding my own version.
- OrderedDictionary class doesn't exist so I would have to write own my and/or borrow an implementation.
- StringComparer.InvariantCultureIgnoreCase doesn't exist although I can just use CurrentCultureIgnoreCase
- Dns.GetHostEntry doesn't exist but I can use the newly added Dns.GetHostName().
- The ConnectionStringSettings class doesn't exist on .NET Core. I would probably have to write my own and/or borrow an implementation in order to parse connection strings.

So there are a number of things that can be done but it will definitely require compiler directives such as:

```csharp
#if !NET40
namespace System.Runtime.Serialization
{
    public class SerializableAttribute : Attribute
    {
    }
}
#endif 
```

#### Roadblocks found
- System.Configuration doesn't exist on .NET Core and is used currently to access connection strings via ```ConfigurationManager.ConnectionStrings[ connectionStringName ]```
- DbProviderFactories class doesn't exist on .NET Core. Discussion for future inclusion is happening on [this GitHub issue](https://github.com/dotnet/corefx/issues/4571).

These roadblocks can be cleared... I just don't currently know how yet without further research. I would most likely look at the other popular micro-ORMs that have already ported.

Other relevant links:
- [Porting to .NET Core](https://blogs.msdn.microsoft.com/dotnet/2016/02/10/porting-to-net-core/)
- [Sharing code across platforms](https://blogs.msdn.microsoft.com/dotnet/2014/04/21/sharing-code-across-platforms/)
- [Porting a .Net Framework Library to .Net Core](http://www.michael-whelan.net/porting-dotnet-framework-library-to-dotnet-core/)

**Comment by westdavidr** [comment]
@randyburden Yeah, I did the same thing and discovered the same things. 

About the connection strings... I was wondering if it would be acceptable to just require the client to supply the connection string instead of Sequelocity trying to come up with it.

I had the same thought about checking out other micro-ORMS and seeing what they do. 

Haha on that DbProviderFactories issue, there's quite a debate going on there. 

And, just because the teams at Ambit aren't actively asking for a .NET Core doesn't mean we can't get ahead of it and have something ready! I'd be happy to contribute in any way I can.
